mesonbuild/build.py

```diff
@@ -440,6 +441,39 @@ def can_compile_remove_sources(compiler, sources):
                 removed = True
         return removed
 
+    def process_compilers_late(self):
+        """Processes additional compilers after kwargs have been evaluated.
+
+        This can add extra compilers that might be required by keyword
+        arguments, such as link_with or dependencies. It will also try to guess
+        which compiler to use if one hasn't been selected already.
+        """
+        # Populate list of compilers
+        if self.is_cross:
+            compilers = self.environment.coredata.cross_compilers
+        else:
+            compilers = self.environment.coredata.compilers
+
+        # If this library is linked against another library we need to consider
+        # the languages of those libraries as well.
+        if self.link_targets or self.link_whole_targets:
+            def iter_compilers():
+                yield from (l.compilers.items() for l in self.link_targets)
+                yield from (l.compilers.items() for l in self.link_whole_targets)
+            self.compilers.update(OrderedDict(
+                sorted((i for i in itertools.chain.from_iterable(iter_compilers())
+                        if i[0] in clike_langs),
+                       key=lambda t: sort_clike(t[0]))))
```

Comment:
Isn't this code a bit too 'smart'? Might be difficult to read for people unfamiliar with how that itertools function works.

I would recommend unravelling it so that it's clearer what it does. Our primary purpose in selecting Python as the implementation language was to make it easier for people to contribute. I would prefer a simpler version of this that's 10 lines longer. :)

Reply:
lol, okay. I'll push an update.